lib/internal/vm/Module.js

```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+const { isContext } = process.binding('contextify');
+const assert = require('assert');
+const errors = require('internal/errors');
+const Loader = require('internal/loader/Loader');
+const { ModuleWrap } = internalBinding('module_wrap');
+
+const {
+  kUninstantiated,
+  kInstantiating,
+  kInstantiated,
+  kEvaluating,
+  kEvaluated,
+  kErrored,
+} = internalBinding('module_wrap');
+
+let globalModuleId = 0;
+const kWrap = Symbol('wrap');
+const kContextModuleCount = Symbol('contextModuleCount');
+const kLoaderContextInjection = '____loaderContextInjection____';
+
+class Module {
+  constructor(src, options = {}) {
+    let context;
+    if (options.context !== undefined && isContext(options.context.sandbox)) {
+      context = options.context.sandbox;
+      if (options.context[kContextModuleCount] === undefined)
+        throw new errors.TypeError('ERR_CONTEXT_NOT_MODULE_CONTEXT');
+    }
+
+    const url = options.url ||
+      `vm:module(${context !== undefined ?
+        options.context[kContextModuleCount]++ :
+        globalModuleId++})`;
+
+    const wrap = new ModuleWrap(src, url, {
+      context,
+      lineOffset: options.lineOffset,
+      columnOffset: options.columnOffset,
+    });
+
+    Object.defineProperties(this, {
+      [kWrap]: { value: wrap },
+      context: { value: options.context },
+    });
+  }
+
+  get status() {
+    return this[kWrap].getStatus();
+  }
+
+  get url() {
+    return this[kWrap].url;
+  }
+
+  get namespace() {
+    assert(this[kWrap].getStatus() === kEvaluated, 'Module has not evaluated');
+    return this[kWrap].namespace();
+  }
+
+  async link(fn) {
```

Comment:
This needs to check that

1. no other linking process is active, and
2. the module's status is still `uninitialized`.

Reply:
i'm fairly sure that ModuleWrap's link and v8's Module already handle that, but i'll look into it

Reply:
Doesn't look like ModuleWrap::Link is handling it, and that method only calls upon `v8::Module::GetModuleRequest()` which doesn't really care about the state of the module in question. So yes, those two things are still outstanding problems.

(Feel free to add these errors/asserts on the C++ side, but on the JavaScript side too for error objects with codes.)